### PR TITLE
Nice database migrations

### PR DESCRIPTION
**Pull request by simonrozsival**
## What's this?
This PR simplifies the way we implement DB migrations and makes it easier to add new migrations in the future.

### Relationships

No directly related issue.

## Why do we want this?

Adding a migration is confusing at the moment, because we have a single method which has several if branches, some of them empty, and somebody who is not familiar with this code might be confused at first and might make a mistake, which would result in corrupting user's data if it was overlooked during code review.

## How is it done?

Migrations, which are independent, will now be separate classes with single responsibility: execute the migration from an old schema to the target schema version.

The `Database` class holds a list of migrations, which should be used, and applies them in a correct order (from lowest target schema versions to the highest).

There might be several classes targeting the same schema, but there must be no dependence between the migrations as the order of them is not guaranteed in any way.

Creating migrations will still be quite challenging and will require a lot of focus (when the schema changes, some of the old migrations might need to be replaced with different ones (e.g., when some property is renamed twice).

### Why not another way?
This is a nice and clean way which should be future-proof.

### Side effects
No side effects.

## Review hints

Try creating your own migration and change some of the `Realm*` classes and bump the schema version, then run the app and use Realm Studio (or the debugger) to see that the migration was applied correctly.

## :squid: Permissions

I will 🦑 this if we agree that it is a good thing to have.

### CODE REVIEW

**Comment by amulware** [comment]
I'm not sure if I like it. What if we change tags again in the future and we need to partially replace the existing migration depending on version numbers? This also makes it very hard to see what migrations we have and what they do. Lastly, we're trying to improve startup performance, not the other way around (even if this will probably be a very tiny difference).

Also, YAGNI.

File: Toggl.PrimeRadiant.Realm/Database.cs [hunk 1]
```diff
@@ -75,30 +82,24 @@ private RealmConfiguration createRealmConfiguration()
...
-                    if (oldSchemaVersion < 6)
-                    {
-                        // nothing needs explicit updating when updating from schema 4 up to 6
-                    }
+        private Action<IMigration> ensureApplicabilityOfMigrationFrom(ulong oldSchemaVersion)
+            => migration =>
+            {
+                if (oldSchemaVersion < migration.MinimumSchemaVersionToMigrateFrom)
+                {
+                    throw new InvalidOperationException(
```
Comment (amulware) [inline Toggl.PrimeRadiant.Realm/Database.cs]: So we'll make it crash for some users at runtime in a pretty untestable way? (We should have tests running all possible migrations, ideally with checking data integrity.)

File: Toggl.PrimeRadiant.Realm/Migrations/TagServerDeletedAt.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+using System.Linq;
+
+namespace Toggl.PrimeRadiant.Realm.Migrations
+{
+    public sealed class TagServerDeletedAt : IMigration
+    {
+        public ulong MinimumSchemaVersionToMigrateFrom { get; } = 0;
+
+        public ulong TargetSchemaVersion { get; } = 4;
```
Comment (amulware) [inline Toggl.PrimeRadiant.Realm/Migrations/TagServerDeletedAt.cs]: `=>` please.

File: Toggl.PrimeRadiant.Realm/Database.cs [hunk 3]
```diff
@@ -12,6 +14,11 @@ public sealed class Database : ITogglDatabase
     {
         private readonly RealmConfiguration realmConfiguration;
 
+        private readonly IMigration[] migrations =
```
Comment (amulware) [inline Toggl.PrimeRadiant.Realm/Database.cs]: Allocating a thing that we will hardly ever use? :/ At least make it lazy? Or better a calculated property that returns a new array only when we're actually migrating.

**Comment by amulware** [comment]
I think the thing we have to optimise for with the migration stuff is (future) correctness. Not ease of writing, and maybe not even ease of reading, if it gets in the way.

**Comment by simonrozsival** [comment]
I can't see any implications on startup performance. Do you to the array of migration instances? That could be lazy loaded, if we want to optimize that.

I would argue that it is easier to see what migrations we have - they now have names (name of the class).

If we need to replace some migration, we simply remove the old class and replace it with a more sophisticated version. That is the same as with the `if-if-if` "pattern", you would also have to rewrite that code.

Ad YAGNI - I just started working on my proposed solution for #1728 to see, if it would help, and adding just a few migrations makes the code quite chaotic (it gets very long).

**Comment by simonrozsival** [comment]
I will close this for now and reopen/introduce again when we need to add some new migrations.

**Pull request closed without merging**